src/target/riscv/riscv.c

```diff
@@ -598,90 +606,174 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
-	tdata1 = set_field(tdata1, CSR_MCONTROL_EXECUTE, trigger->execute);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_LOAD, trigger->read);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_STORE, trigger->write);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_ANY & 3);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_ANY >> 2) & 3);
-
-	if (trigger->execute || trigger->length == 1)
-		goto MATCH_EQUAL;
-	if (!can_use_napot_match(trigger, &tdata2))
-		goto MATCH_GE_LT;
-
-	unsigned int idx;
-	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, false, &idx);
-	if (ret != ERROR_OK)
-		return ret;
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_NAPOT);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	ret = set_trigger(target, idx, tdata1, tdata2, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-	if (ret != ERROR_OK) {
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_GE_LT; /* Fallback to chained MATCH using GT and LE */
-		return ret;
+	for (unsigned int idx = 0;
+		 find_next_trigger(target, trigger_type, false, &idx) == ERROR_OK &&
+		 ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+		 ++idx) {
+		ret = set_trigger(target, idx, trig_info.tdata1, trig_info.tdata2,
+						  trig_info.tdata1_ignore_mask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			return ERROR_OK;
+		}
 	}
-	r->trigger_unique_id[idx] = trigger->unique_id;
-	return ERROR_OK;
+	return ret;
+}
 
-MATCH_GE_LT:
-	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, true, &idx);
-	if (ret != ERROR_OK)
-		return ret;
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_GE);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_ENABLED);
-	tdata2 = trigger->address;
-	ret = set_trigger(target, idx, tdata1, tdata2, 0);
-	if (ret != ERROR_OK) {
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
-		return ret;
+static int
+try_chained_trigger_request_match(struct target *target,
+								  struct trigger *trigger, int trigger_type,
+								  struct trigger_request_info trig1_info,
+								  struct trigger_request_info trig2_info)
+{
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+	RISCV_INFO(r);
+
+	for (unsigned int idx = 0;
+		 find_next_trigger(target, trigger_type, true, &idx) == ERROR_OK &&
+		 ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+		 ++idx) {
+		ret = set_trigger(target, idx, trig1_info.tdata1, trig1_info.tdata2,
+						  trig1_info.tdata1_ignore_mask);
+		if (ret != ERROR_OK)
+			continue;
+		ret = set_trigger(target, idx + 1, trig2_info.tdata1, trig2_info.tdata2,
+						  trig2_info.tdata1_ignore_mask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			r->trigger_unique_id[idx + 1] = trigger->unique_id;
+			return ERROR_OK;
+		} else
+			set_trigger(target, idx, 0, 0,
+						0); /* Undo the setting of the previous trigger*/
```

Comment:
You can reformat it like this:

```suggestion
			/* Undo the setting of the previous trigger */
			set_trigger(target, idx, 0, 0, 0); 
```

Reply:
Addressed.